Mathlib/NumberTheory/Divisors.lean

```diff
@@ -514,3 +530,114 @@ theorem disjoint_divisors_filter_isPrimePow {a b : ℕ} (hab : a.Coprime b) :
   exact hn.ne_one (Nat.eq_one_of_dvd_coprimes hab han hbn)
 
 end Nat
+
+namespace Int
+variable {xy : ℤ × ℤ} {x y z : ℤ}
+
+-- Local notation for the embeddings `n ↦ n, n ↦ -n : ℕ → ℤ`
+local notation "natCast" => Nat.castEmbedding (R := ℤ)
+local notation "negNatCast" =>
+  Function.Embedding.trans Nat.castEmbedding (Equiv.toEmbedding (Equiv.neg ℤ))
+
+/-- Pairs of divisors of an integer as a finset.
+
+`z.divisorsAntidiag` is the finset of pairs `(a, b) : ℤ × ℤ` such that `a * b = z`.
+By convention, we set `Int.divisorsAntidiag 0 = ∅`.
+
+O(|z|). Computed from `Nat.divisorsAntidiagonal`. -/
+def divisorsAntidiag : (z : ℤ) → Finset (ℤ × ℤ)
+  | (n : ℕ) =>
+    let s : Finset (ℕ × ℕ) := n.divisorsAntidiagonal
+    (s.map <| .prodMap natCast natCast).disjUnion (s.map <| .prodMap negNatCast negNatCast) <| by
+      simp +contextual [s, disjoint_left, eq_comm]
+  | negSucc n =>
+    let s : Finset (ℕ × ℕ) := (n + 1).divisorsAntidiagonal
+    (s.map <| .prodMap natCast negNatCast).disjUnion (s.map <| .prodMap negNatCast natCast) <| by
+      simp +contextual [s, disjoint_left, eq_comm, forall_swap (α := _ * _ = _)]
+
+@[simp]
+lemma mem_divisorsAntidiag :
+    ∀ {z} {xy : ℤ × ℤ}, xy ∈ divisorsAntidiag z ↔ xy.fst * xy.snd = z ∧ z ≠ 0
+  | (n : ℕ), ((x : ℕ), (y : ℕ)) => by
+    simp [divisorsAntidiag]
+    norm_cast
+    simp +contextual [eq_comm]
+  | (n : ℕ), (negSucc x, negSucc y) => by
+    simp [divisorsAntidiag, negSucc_eq, -neg_add_rev]
+    norm_cast
+    simp +contextual [eq_comm]
+  | (n : ℕ), ((x : ℕ), negSucc y) => by
+    simp [divisorsAntidiag, negSucc_eq, -neg_add_rev]
+    norm_cast
+    aesop
+  | (n : ℕ), (negSucc x, (y : ℕ)) => by
+    simp [divisorsAntidiag]
+    norm_cast
+    aesop
+  | .negSucc n, ((x : ℕ), (y : ℕ)) => by
+    simp [divisorsAntidiag]
+    norm_cast
+    aesop
+  | .negSucc n, (negSucc x, negSucc y) => by
+    simp [divisorsAntidiag, negSucc_eq, -neg_add_rev]
+    norm_cast
+    simp +contextual [eq_comm]
+  | .negSucc n, ((x : ℕ), negSucc y) => by
+    simp [divisorsAntidiag, negSucc_eq, -neg_add_rev]
+    norm_cast
+    aesop
+  | .negSucc n, (negSucc x, (y : ℕ)) => by
+    simp [divisorsAntidiag, negSucc_eq, -neg_add_rev]
+    norm_cast
+    simp +contextual [eq_comm]
+
+@[simp] lemma divisorsAntidiag_zero : divisorsAntidiag 0 = ∅ := rfl
+
+lemma prodMk_mem_divisorsAntidiag (hz : z ≠ 0) : (x, y) ∈ z.divisorsAntidiag ↔ x * y = z := by
+  simp [hz]
+
+-- The left hand side is not in simp normal form, see the variant below.
+lemma swap_mem_divisorsAntidiag : xy.swap ∈ z.divisorsAntidiag ↔ xy ∈ z.divisorsAntidiag := by
+  simp [mul_comm]
+
+/-- `Int.swap_mem_divisorsAntidiag` with the LHS in simp normal form. -/
+@[simp]
+lemma swap_mem_divisorsAntidiag_simpNF :
+    xy.snd * xy.fst = z ∧ ¬z = 0 ↔ xy ∈ z.divisorsAntidiag := by simp [mul_comm]
```

Comment:
Why do we even care about this lemma existing? The RHS is not simp-normal form either.

I suspect it would be easier to just put `@[simp high]` on `swap_mem_divisorsAntidiag` and be done with it.

Reply:
I tend to agree. Would be good to acquire a guideline here